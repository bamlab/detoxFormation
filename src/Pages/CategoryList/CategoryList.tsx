import React, { FunctionComponent, useState, useEffect } from 'react';
import { ScrollView, StyleSheet, Text } from 'react-native';

import { Category } from '../../Components/Category/Category';

export const API_BASE_URL = 'https://www.themealdb.com/api/json/v1/1';

type RawCategory = {
  idCategory?: string;
  strCategory?: string;
  strCategoryThumb?: string;
  strCategoryDescription?: string;
};

export const CategoryList: FunctionComponent = () => {
  const [categories, setCategories] = useState<RawCategory[]>([]);
  const [isLoading, setLoading] = useState(true);

  const getCategories = async () => {
    try {
      const response = await fetch(`${API_BASE_URL}/categories.php`);
      const json: { categories: RawCategory[] } = await response.json();
      if (json.categories) {
        setCategories(json.categories);
      }
    } catch (error) {
      console.log(error);
    }
    setLoading(false);
  };

  useEffect(() => {
    void getCategories();
  }, []);

  return (
    <ScrollView>
      <Text style={styles.title}>The Meal App</Text>
      {isLoading ? (
        <Text>Loading...</Text>
      ) : (
        categories.map(category => {
          const { strCategory, strCategoryThumb } = category;

          if (strCategory && strCategoryThumb) {
            return (
              <Category
                key={category.idCategory}
                category={{ strCategory, strCategoryThumb }}
              />
            );
          }
          return null;
        })
      )}
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  title: {
    fontSize: 40,
    fontWeight: 'bold',
    padding: 10,
  },
});
