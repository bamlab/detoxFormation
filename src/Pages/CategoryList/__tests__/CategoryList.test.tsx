import 'react-native';
import React from 'react';
import {
  render,
  waitForElementToBeRemoved,
} from '@testing-library/react-native';
import nock from 'nock';

import { API_BASE_URL, CategoryList } from '../CategoryList';
import categoryFixture from './category.fixture.json';

describe('[Page] CategoryList', () => {
  test('renders correctly with correct api data', async () => {
    nock(API_BASE_URL).get('/categories.php').reply(200, categoryFixture);

    const component = render(<CategoryList />);
    await waitForElementToBeRemoved(() => component.getByText('Loading...'));

    expect(component).toMatchSnapshot();
  });

  test('renders correctly with some undefined value', async () => {
    nock(API_BASE_URL)
      .get('/categories.php')
      .reply(200, {
        categories: [
          {
            idCategory: '1',
            strCategory: undefined,
            strCategoryThumb:
              'https://www.themealdb.com/images/category/beef.png',
            strCategoryDescription:
              'Beef is the culinary name for meat from cattle, particularly skeletal muscle. Humans have been eating beef since prehistoric times.[1] Beef is a source of high-quality protein and essential nutrients.[2]',
          },
          {
            idCategory: '2',
            strCategory: 'Chicken',
            strCategoryThumb:
              'https://www.themealdb.com/images/category/chicken.png',
            strCategoryDescription:
              'Chicken is a type of domesticated fowl, a subspecies of the red junglefowl. It is one of the most common and widespread domestic animals, with a total population of more than 19 billion as of 2011.[1] Humans commonly keep chickens as a source of food (consuming both their meat and eggs) and, more rarely, as pets.',
          },
        ],
      });

    const component = render(<CategoryList />);
    await waitForElementToBeRemoved(() => component.getByText('Loading...'));

    expect(component).toMatchSnapshot();
  });

  test('renders correctly with an undefined/null response from the api', async () => {
    nock(API_BASE_URL).get('/categories.php').reply(200, undefined);

    const component = render(<CategoryList />);
    await waitForElementToBeRemoved(() => component.getByText('Loading...'));

    expect(component).toMatchSnapshot();
  });
});
