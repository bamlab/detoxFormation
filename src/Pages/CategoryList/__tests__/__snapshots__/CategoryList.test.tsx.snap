// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Page] CategoryList renders correctly with an undefined/null response from the api 1`] = `
<RCTScrollView
  testID="scroll-view"
>
  <View>
    <Text
      style={
        Object {
          "fontSize": 40,
          "fontWeight": "bold",
          "padding": 10,
        }
      }
    >
      The Meal App
    </Text>
  </View>
</RCTScrollView>
`;

exports[`[Page] CategoryList renders correctly with correct api data 1`] = `
<RCTScrollView
  testID="scroll-view"
>
  <View>
    <Text
      style={
        Object {
          "fontSize": 40,
          "fontWeight": "bold",
          "padding": 10,
        }
      }
    >
      The Meal App
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Beef
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/beef.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Chicken
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/chicken.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Dessert
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/dessert.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Lamb
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/lamb.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Miscellaneous
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/miscellaneous.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Pasta
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/pasta.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Pork
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/pork.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Seafood
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/seafood.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Side
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/side.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Starter
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/starter.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Vegan
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/vegan.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Vegetarian
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/vegetarian.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Breakfast
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/breakfast.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Goat
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/goat.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;

exports[`[Page] CategoryList renders correctly with some undefined value 1`] = `
<RCTScrollView
  testID="scroll-view"
>
  <View>
    <Text
      style={
        Object {
          "fontSize": 40,
          "fontWeight": "bold",
          "padding": 10,
        }
      }
    >
      The Meal App
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#FFF8",
            "color": "black",
            "fontSize": 20,
            "marginLeft": 20,
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        Chicken
      </Text>
      <Image
        source={
          Object {
            "uri": "https://www.themealdb.com/images/category/chicken.png",
          }
        }
        style={
          Object {
            "height": 75,
            "width": "100%",
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;
